### PR TITLE
tests/kernel/smp: improve arch_sched_ipi test

### PR DESCRIPTION
**Pull request by npitre**
Make sure an IPI is raised only once per each CPU except for the
CPU that invoked it. Avoid any scheduling points during the test so
not to trigger unwanted extra IPIs.

### CODE REVIEW

File: tests/kernel/smp/src/main.c [hunk 1]
```diff
@@ -652,23 +654,32 @@ ZTEST(smp, test_smp_ipi)
...
+		atomic_clear(&sched_ipi_has_called);
+		atomic_clear(sched_ipi_on_cpu);
 		arch_sched_ipi();
 
-		/* Need to wait longer than we think, loaded CI
-		 * systems need to wait for host scheduling to run the
-		 * other CPU's thread.
-		 */
-		k_msleep(100);
+		k_busy_wait(DELAY_US);
```
Comment (andyross) [inline tests/kernel/smp/src/main.c:668]: This doesn't look right.  The sense of that comment was that in CI environments[1],  IPI delivery is going to depend on scheduling on the host processors, which needless to say aren't going to be able to deliver within a microsecond-scale delay.  (FWIW, that comment sounds like my writing, but I have zero memory of this particular issue.  I don't think I wrote the IPI test originally though.)

Also k_busy_wait() tends to bound on local interrupt locks on most architectures and seems likely to make interrupt latency tests like this **worse** and not better.  I guess I don't see the value here?

[1]  Where the host is loaded by twister and SMP cores, which run in separate host threads, are going to see "time warp" effects relative to each other as they get scheduled in and out in competition with other build/test processes.

**Review by andyross — CHANGES_REQUESTED** [review]
Nitpick on an edge case that seems likely to have regressed.  More generally I guess I don't see the point here?  I get that there's an aesthetic desire to avoid incidental IPIs during the IPI test, but that's not really possible here in a test that has already run **many** threads across CPUs just to get as far as the start of the test case.

If we want a unit test of the lowest level primitives like this, it probably wants to be done with CONFIG_SMP=n and be written to the lower level MP API.   There's such a test already in tests/kernel/mp, which is a vestigial remnant of the original SMP bringup work.  I don't know how many architectures have IPI implementations that work in that config though.

Mild -1.  Mostly just because it seems like churn, and churn in tests is usually bad unless it clearly expands coverage.  Feel free to override if there's a need for this to go in.

File: tests/kernel/smp/src/main.c (hunk 1 above)
Comment (npitre) [reply tests/kernel/smp/src/main.c:668]: On the contrary I think it is the original comment that is wrong. K_msleep() does trigger extra scheduler IPIs in addition to the explicit one here. Furthermore, this test does use this k_busy_wait() all over the place to give other CPUs the time to schedule and execute newly created threads, etc. So if this was a problem the entire test would be broken.
As it is, this IPI test is a bit pointless. We want to help new architecture implementors to do it right i.e. raise an IRQ on all CPUs but the current one and that is not tested anywhere. Given that the modification to the test do  work on all supported architectures already makes me wonder why you appear to be so concerned?

**Comment by npitre** [comment]
Incidentally this improved test just saved me from a bad IPI 
implementation on RISC-V in the case where the hartid doesn't match the 
Zephyr CPU number. It is good to not raise an IRQ on the sending CPU as 
long as the excluded C^PU is actually the right one physically. So this 
proved to be useful already.

**Pull request closed without merging**